pennylane/compiler/qjit_api.py

```diff
@@ -105,8 +107,177 @@ def circuit(theta):
     """
 
     if not available(compiler):
-        raise RuntimeError(f"The {compiler} package is not installed.")
+        raise CompileError(f"The {compiler} package is not installed.")
 
     compilers = AvailableCompilers.names_entrypoints
     qjit_loader = compilers[compiler]["qjit"].load()
     return qjit_loader(fn=fn, *args, **kwargs)
+
+
+def while_loop(*args, compiler="catalyst", **kwargs):
+    """A :func:`~.qjit` compatible while-loop for PennyLane programs.
+
+    .. note::
+
+        When used with :func:`~.qjit`, this function only supports
+        the Catalyst compiler. Please see
+        :func:`catalyst.while_loop` for more details.
+
+        Please see the Catalyst :doc:`quickstart guide <catalyst:dev/quick_start>`,
+        as well as the :doc:`sharp bits and debugging tips <catalyst:dev/sharp_bits>`
+        page for an overview of the differences between Catalyst and PennyLane.
+
+    This decorator provides a functional version of the traditional while
+    loop, similar to ``jax.lax.while_loop``. That is, any variables that are
+    modified across iterations need to be provided as inputs and outputs to
+    the loop body function:
+
+    - Input arguments contain the value of a variable at the start of an
+      iteration
+
+    - Output arguments contain the value at the end of the iteration. The
+      outputs are then fed back as inputs to the next iteration.
+
+    The final iteration values are also returned from the
+    transformed function.
+
+    The semantics of ``while_loop`` are given by the following Python pseudo-code:
+
+    .. code-block:: python
+
+        def while_loop(cond_fun, body_fun, *args):
+            while cond_fun(*args):
+                args = body_fn(*args)
+            return args
+
+    Args:
+        cond_fn (Callable): the condition function in the while loop
+        compiler(str): name of the compiler package (default value is ``catalyst``)
+
+    Returns:
+        Callable: A wrapper around the while-loop function.
+
+    Raises:
+        CompileError: if the compiler is not installed
+
+    **Example**
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", wires=1)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def circuit(x: float):
+
+            @qml.while_loop(lambda x: x < 2.0)
+            def loop_rx(x):
+                # perform some work and update (some of) the arguments
+                qml.RX(x, wires=0)
+                return x ** 2
+
+            # apply the while loop
+            final_x = loop_rx(x)
+
+            return qml.expval(qml.PauliZ(0)), final_x
+
+    >>> circuit(1.6)
+    [array(-0.02919952), array(2.56)]
+    """
```

Comment:
I foresee great sadness trying to keep these docstrings up to date 😢

Reply:
@dime10 this can be my job :)

Reply:
:cry: